### ISSUE TITLE
How to serialize Submission object?

### ISSUE THREAD

**Issue by Saketme** [thatJavaNerd/JRAW#176]
I'd like to save my `Submission` object (without comments) across Activity recreations so that I can restore them later but it neither implements Parcelable nor Serializable. Is there any supported way I can do this apart from manual serialization?

Update: So I'm doing this:

```
@Override
    public void onSaveInstanceState(Bundle outState) {
        if (activeSubmission != null) {
            JsonNode dataNode = activeSubmission.getDataNode();
            outState.putString(KEY_SUBMISSION_JSON, dataNode.toString());
        }
        super.onSaveInstanceState(outState);
    }

    private void onRestoreSavedInstanceState(Bundle savedInstanceState) {
        if (savedInstanceState.containsKey(KEY_SUBMISSION_JSON)) {
            try {
                String submissionJson = savedInstanceState.getString(KEY_SUBMISSION_JSON);
                JsonNode jsonNode = new ObjectMapper().readTree(submissionJson);
                populateUi(new Submission(jsonNode));

            } catch (IOException e) {
                Timber.e(e, "Couldn't deserialize Submission for state restoration");
            }
        }
    }
```

Do you have any feedback on this?

**Comment by vishnumad** [comment]
You could use [EventBust](https://github.com/greenrobot/EventBus) to pass objects between activities.

**Comment by ccrama** [comment]
You can see how I implemented it in Slide [here](https://github.com/ccrama/Slide/blob/master/app/src/main/java/me/ccrama/redditslide/OfflineSubreddit.java#L62). Basically, you get the JsonNode object from the Submission, turn it into a string, and vice versa when you are re-creating it. 

If you are starting from scratch/want a more reactive approach, one thing I would recommend is creating a wrapper with integration with Realm, which is what I decided to do in my Slide port for iOS.

**Comment by Saketme** [comment]
@ccrama Yeah, even I ended up doing the same thing. I did notice that jackson is quite complicated. After working with gson/moshi, it feels like a monolith. Maybe I'm wrong.

About Realm, the fact that it does not support multithreading is a big bummer for me. I've worked with realm in the past and we ended up with workarounds around this. I plan to instead add observables to my SQL database. Any other advantages of using realm?

**Comment by fbis251** [comment]
@Saketme It's not true that realm doesn't support multiple threads, in fact it was designed to be easy to work with threading in mind due to its zero copy approach to storage. I recommend you read about Realm's design here: [Designing a Database: Realm Threading Deep Dive](https://realm.io/news/threading-deep-dive/)

For dealing with threading on Android specifically, the Realm manual provides some examples, check out [Working With Android](https://realm.io/docs/java/latest/#working-with-android)'s AsyncTask example

Basically when dealing with threads, each thread should get its own realm instance, perform its insert/update/delete operations, then the realm instance is closed. That is *per thread*.

By using a change listener you can then update the UI with the newly updated data in your Realm in a completely separate thread

**Comment by Saketme** [comment]
> in fact it was designed to be easy to work with threading in mind due to its zero copy approach to storage.

The problem was that I was constantly switching threads through-out the app. For instance, I'd fetch a list of items on the main thread (because reading from Realm is so fast), but for making changes to them, I'd prefer doing that in a background thread. And since Realm doesn't allow you to pass around objects, I was forced to save the IDs of the objects somewhere and read them again in the other thread. 

When working with Rx where managing asynchronous tasks is a breeze, I started hating Realm for this.

> By using a change listener you can then update the UI with the newly updated data in your Realm in a completely separate thread

So by subscribing to a DB for updates solves this thing for me, removing the need of Realm's always-updated feature. What else does Realm offer?
